### PR TITLE
Get more integration test to work

### PR DESCRIPTION
**Pull request by beeender**
- Add a new TestRule RunWithRemoteService to allow a test case create a
  service running in a remote process. This is very much like what we have
  for the RealmInterProcessTests but with more flexibility, eg.:
  Separate service for separate test case.
- Enabled sync interaction tests: notify by simple commit, notify by
  lots of commits.
- Deprecate RemoteProcessService.

### CODE REVIEW

**Comment by beeender** [comment]
This works great with running a signal test, but fails when running two or more ...
We need to figure out a way to peacefully restart sync service and client.

File: realm/realm-library/src/androidTest/java/io/realm/rule/RunWithRemoteService.java [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+import static junit.framework.Assert.assertTrue;
+import static junit.framework.Assert.fail;
+
+/**
+ * This is a helper {@link TestRule} to do test which needs interaction with a remote process.
+ * To use this:
+ * 1. Define a subclass of {@link RemoteTestService} and create steps as static member of it. Those steps should be
+ * named as "stepA_doXXX", "stepB_doYYY", etc. to indicate the order of them.
+ * 2. Add the service into the AndroidManifest.xml. And the android:process property must be ":remote".
+ * 3. Annotated your test case by {@link RunTestWithRemoteService} with your remote service class.
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/rule/RunWithRemoteService.java]: Annotated -> Annotate

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/HttpUtils.java [hunk 2]
```diff
@@ -38,6 +38,16 @@
     // will forward this query to the host, running the integration test server on 8888
     private final static String START_SERVER = "http://127.0.0.1:8888/start";
     private final static String STOP_SERVER = "http://127.0.0.1:8888/stop";
+    private static boolean inited = false;
+
+    // Start the ROS service only once since there are problems with reseting it. Also it makes testing faster.
```
Comment (cmelchior) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/HttpUtils.java]: Can you elaborate on the problems resetting it? 
This pattern here seems a bit dangerous as it will leave the server running after the last integration test is done. We can perhaps live with that for now though.

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/HttpUtils.java [hunk 3]
```diff
@@ -38,6 +38,16 @@
...
+    public static void init() throws Exception {
```
Comment (cmelchior) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/HttpUtils.java]: This should be synchronized

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/UserFactory.java [hunk 4]
```diff
@@ -16,14 +16,38 @@
...
+import io.realm.log.RealmLog;
 
 // Must be in `io.realm.objectserver` to work around package protected methods.
 public class UserFactory {
+    private static final String PASSWORD = "myPassw0rd";
+    // Since the integration tests need use the same user for different processes, we create a new user name when the
+    // test starts and store it in a Realm. Then it can be retrieved for every process.
+    private String userName;
+    private static UserFactory instance;
+    private static RealmConfiguration configuration = new RealmConfiguration.Builder()
```
Comment (cmelchior) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/UserFactory.java]: This require `Realm.init()` to be called before using this class.  Perhaps we should add a comment to the header about that?

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/UserFactory.java [hunk 5]
```diff
@@ -32,4 +56,43 @@ public static SyncUser createAdminUser(String authUrl) {
         SyncCredentials credentials = SyncCredentials.custom("admin", "debug", null);
         return SyncUser.login(credentials, authUrl);
     }
+
+    // Since we don't have a reliable way to reset the sync server and client, just use a new user for every test case.
+    public static void resetInstance() {
+        instance = null;
+        SecureRandom random = new SecureRandom();
```
Comment (cmelchior) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/UserFactory.java]: Just use `UUID.random().toString()` as user name, it should be safe enough.

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/UserFactory.java [hunk 6]
```diff
@@ -32,4 +56,43 @@ public static SyncUser createAdminUser(String authUrl) {
         SyncCredentials credentials = SyncCredentials.custom("admin", "debug", null);
         return SyncUser.login(credentials, authUrl);
     }
+
+    // Since we don't have a reliable way to reset the sync server and client, just use a new user for every test case.
+    public static void resetInstance() {
```
Comment (cmelchior) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/UserFactory.java]: There seems to be a mismatch between the name and description. Perhaps `generateNewUser()` instead?

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/UserFactory.java [hunk 7]
```diff
@@ -32,4 +56,43 @@ public static SyncUser createAdminUser(String authUrl) {
...
+    // The @Before method will be called before the looper tests finished. We need to find a better place to call this.
+    public static void clearInstance()  {
+        Realm realm = Realm.getInstance(configuration);
+        realm.beginTransaction();
+        realm.delete(UserFactoryStore.class);
+        realm.commitTransaction();
+        realm.close();
+    }
+
+    public static UserFactory getInstance() {
```
Comment (cmelchior) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/UserFactory.java]: Needs to be synchronized

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/HttpUtils.java (hunk 2 above)
Comment (beeender) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/HttpUtils.java]: It might be my fault ... After clean up everything, the original solution (start/stop in BeforeClass/AfterClass) works. Will revert this change.

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/HttpUtils.java (hunk 3 above)
Comment (beeender) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/HttpUtils.java]: will remove this.

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/UserFactory.java (hunk 6 above)
Comment (beeender) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/UserFactory.java]: This is to match the method name `getInstance()`. Fixed the comment.

**Comment by beeender** [comment]
Finally got this to work. @realm/java review please.

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProcessCommitTests.java [hunk 8]
```diff
@@ -16,159 +16,202 @@
...
-                    Looper.loop();
-
-                } catch (Throwable e) {
-                    exception[0] = e;
-                    testFinished.countDown();
-                }
+            public void onChange(RealmResults<ProcessInfo> element) {
+                assertEquals(1, all.size());
+                assertEquals("Background_Process1", all.get(0).getName());
+                realm.close();
```
Comment (cmelchior) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProcessCommitTests.java]: Can these clean up methods somehow be moved to the Rule? These tests are so complex that anything we can do to reduce it would be nice.

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProcessCommitTests.java [hunk 9]
```diff
@@ -16,159 +16,202 @@
...
+            @Override
+            protected void run() {
+                user = UserFactory.getInstance().loginWithDefaultUser(Constants.AUTH_URL);
+                String realmUrl = Constants.SYNC_SERVER_URL;
+
+                final SyncConfiguration syncConfig = new SyncConfiguration.Builder(user, realmUrl)
+                        .directory(getService().getRoot())
+                        .name(UUID.randomUUID().toString() + ".realm")
+                        .build();
+                getService().realm = Realm.getInstance(syncConfig);
```
Comment (cmelchior) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProcessCommitTests.java]: I think @bmeike would prefer if this was `setRealm()` and getRealm()` instead :)

File: realm/realm-library/src/androidTest/java/io/realm/rule/RunWithRemoteService.java [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+            super(looper);
+            localMessenger = new Messenger(this);
+        }
+
+        @Override
+        public void handleMessage(Message msg) {
+            Bundle bundle = msg.getData();
+            String error = bundle.getString(RemoteTestService.BUNDLE_KEY_ERROR);
+            if (error != null) {
+                // Assert and show error from value process
```
Comment (beeender) [inline realm/realm-library/src/androidTest/java/io/realm/rule/RunWithRemoteService.java]: I don't understand either, probably i wanted to say remote process.

File: realm/realm-library/src/androidTest/java/io/realm/services/RemoteProcessService.java [hunk 11]
```diff
@@ -34,6 +34,7 @@
 
 /**
  * Helper service for multi-processes support testing.
+ * @deprecated use {@link RemoteTestService} instead.
```
Comment (beeender) [inline realm/realm-library/src/androidTest/java/io/realm/services/RemoteProcessService.java]: It is still used in `RealmInterprocessTest` which I don't have time to refactor yet.

File: realm/realm-library/src/androidTest/java/io/realm/services/RemoteTestService.java [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.HashMap;
+import java.util.Map;
+
+import io.realm.Realm;
+import io.realm.log.RealmLog;
+
+/**
+ * Helper class for multi-processes support testing.
```
Comment (beeender) [inline realm/realm-library/src/androidTest/java/io/realm/services/RemoteTestService.java]: Description in `RunWithRemoteService`. Added preferred naming rules there, and added a link here.

File: realm/realm-library/src/androidTest/java/io/realm/services/RemoteTestService.java [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+
+        protected Step(int base, int id) {
+            this.message = base + id;
+            stepMap.put(this.message, this);
+        }
+
+        protected abstract void run();
+
+        protected RemoteTestService getService() {
+            return RemoteTestService.thiz;
```
Comment (beeender) [inline realm/realm-library/src/androidTest/java/io/realm/services/RemoteTestService.java]: This is a method of `Step` which is a static inner class.

File: realm/realm-library/src/androidTest/java/io/realm/services/RemoteTestService.java [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+            } catch (RemoteException e) {
+                RealmLog.error(e);
+            }
+        }
+    }
+
+    public static final String BUNDLE_KEY_ERROR = "error";
+    @SuppressLint("UseSparseArrays")
+    private static Map<Integer, Step> stepMap = new HashMap<Integer, Step>();
+    public static RemoteTestService thiz;
```
Comment (beeender) [inline realm/realm-library/src/androidTest/java/io/realm/services/RemoteTestService.java]: All the `Step` are defined as static method in the child classes for a reason which I cannot remember... looks very much because of my comment here:
>     // There is no easy way to dynamically ensure step IDs have same value for different processes. So, use the stupid
>    // way.

And the step instance still need to access the `RemoteTestService` instance sometimes.

File: realm/realm-library/src/androidTest/java/io/realm/services/RemoteTestService.java [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+
+    @Override
+    public IBinder onBind(Intent intent) {
+        return messenger.getBinder();
+    }
+
+    @Override
+    public boolean onUnbind(Intent intent) {
+        stopSelf();
+        recursiveDelete(rootFolder);
```
Comment (beeender) [inline realm/realm-library/src/androidTest/java/io/realm/services/RemoteTestService.java]: I think it doesn't, doc from `createTempFile()`:
> This method provides only part of a temporary-file facility. To arrange for a file created by this method to be deleted automatically, use the deleteOnExit() method.
> 
> The prefix argument must be at least three characters long. It is recommended that the prefix be a short, meaningful string such as "hjb" or "mail". The suffix argument may be null, in which case the suffix ".tmp" will be used. 

Also, the `TemporaryFolder` is doing the same.

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProcessCommitTests.java [hunk 16]
```diff
@@ -16,159 +16,202 @@
...
+                Realm realm = getService().realm;
+
+                realm.beginTransaction();
+                ProcessInfo processInfo = realm.createObject(ProcessInfo.class);
+                processInfo.setName("Background_Process1");
+                processInfo.setPid(android.os.Process.myPid());
+                processInfo.setThreadId(Thread.currentThread().getId());
+                realm.commitTransaction();
+                // FIXME: If we close the Realm here, the data won't be able to synced to the main process. Is it a bug
```
Comment (beeender) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProcessCommitTests.java]: Can we get a `SyncSession` instance without any Realm instance opened?

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProcessCommitTests.java (hunk 8 above)
Comment (beeender) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProcessCommitTests.java]: I think it could be achieved by inherit the `RunInLooperThread` which would do the user logout when clean up. I can do it in a following PR.

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/UserFactory.java [hunk 17]
```diff
@@ -16,18 +16,44 @@
...
 
+import java.util.UUID;
+
+import io.realm.Realm;
+import io.realm.RealmConfiguration;
 import io.realm.SyncCredentials;
 import io.realm.SyncUser;
+import io.realm.log.RealmLog;
 
+// Helper class to retrieve users with same IDs even in multi-processes.
```
Comment (beeender) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/UserFactory.java]: Should this store the user? `SyncUser.currentUser()` can get the current login user I think?

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProcessCommitTests.java (hunk 16 above)
Comment (cmelchior) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProcessCommitTests.java]: Not currently, no. This is also something we need to get fixed, because other features depend on this being possible

File: realm/realm-library/src/androidTest/java/io/realm/rule/RunWithRemoteService.java [hunk 18]
```diff
@@ -170,7 +174,7 @@ public void triggerServiceStep(RemoteTestService.Step step) {
         ActivityManager manager = (ActivityManager)getContext().getSystemService(Context.ACTIVITY_SERVICE);
         List<ActivityManager.RunningAppProcessInfo> processInfoList = manager.getRunningAppProcesses();
         for (ActivityManager.RunningAppProcessInfo info : processInfoList) {
-            if (info.processName.equals(getContext().getPackageName() + ":remote")) {
+            if (info.processName.equals(getContext().getPackageName() + REMOTE_PROCESS_POSTFIX)) {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/rule/RunWithRemoteService.java]: I think the word you are looking for is SUFFIX

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProcessCommitTests.java [hunk 19]
```diff
@@ -16,159 +16,207 @@
...
+            @Override
+            protected void run() {
+                getService().getRealm().close();
+                user.logout();
+            }
+        };
+    }
+
+    // 1. Open a sync Realm and listen to changes.
+    // A. Open the same sync Realm and add on object.
```
Comment (nhachicha) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProcessCommitTests.java]: add one object

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProcessCommitTests.java [hunk 20]
```diff
@@ -16,159 +16,207 @@
...
-                    testFinished.countDown();
+            public void onChange(RealmResults<TestObject> element) {
+                int counter = listenerCalledCounter.incrementAndGet();
+                int size = all.size();
+                if (size == 0) {
+                    listenerCalledCounter.decrementAndGet();
+                    return;
+                }
+                assertEquals(0, size % 100); // Added 100 objects every time.
+                assertEquals(counter * 100 -1, all.last().getIntProp());
```
Comment (nhachicha) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProcessCommitTests.java]: space after `-`

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProcessCommitTests.java [hunk 21]
```diff
@@ -16,159 +16,207 @@
...
+                assertEquals("Str" + (counter * 100 -1), all.last().getStringProp());
```
Comment (nhachicha) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProcessCommitTests.java]: space after `-`

File: realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/UserFactory.java [hunk 22]
```diff
@@ -16,18 +16,44 @@
...
 // Must be in `io.realm.objectserver` to work around package protected methods.
+// This require Realm.init() to be called before using this class.
 public class UserFactory {
+    private static final String PASSWORD = "myPassw0rd";
+    // Since the integration tests need use the same user for different processes, we create a new user name when the
```
Comment (nhachicha) [inline realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/utils/UserFactory.java]: need to use
